src/repositories/updater.rs

```diff
@@ -0,0 +1,276 @@
+                 FROM releases
+                 INNER JOIN crates ON (crates.id = releases.crate_id)
+                 WHERE repository_id IS NULL AND repository_url LIKE $1;",
+                &[&format!("%{}%", host)],
+            )?;
+
+            let mut missing_urls = HashSet::new();
+            for row in &needs_backfilling {
+                let id: i32 = row.get("id");
+                let name: String = row.get("name");
+                let version: String = row.get("version");
+                let url: String = row.get("repository_url");
+
+                if missing_urls.contains(&url) {
+                    debug!("{} {} points to a known missing repo", name, version);
+                } else if let Some(node_id) = self.load_repository(&mut conn, &url)? {
+                    conn.execute(
+                        "UPDATE releases SET repository_id = $1 WHERE id = $2;",
+                        &[&node_id, &id],
+                    )?;
+                    info!(
+                        "backfilled {} repository for {} {}",
+                        Self::name(),
+                        name,
+                        version
+                    );
+                } else {
+                    debug!(
+                        "{} {} does not point to a {} repository",
+                        Self::name(),
+                        name,
+                        version
+                    );
+                    missing_urls.insert(url);
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn repository_name(url: &str) -> Option<RepositoryName> {
+        static RE: Lazy<Regex> = Lazy::new(|| {
+            Regex::new(r"https?://(?P<host>.+)/(?P<owner>[\w\._-]+)/(?P<repo>[\w\._-]+)").unwrap()
+        });
+
+        let cap = RE.captures(url)?;
+        let host = cap.name("host").expect("missing group 'host'").as_str();
+        if !Self::hosts().iter().any(|s| *s == host) {
+            return None;
+        }
+        let owner = cap.name("owner").expect("missing group 'owner'").as_str();
+        let repo = cap.name("repo").expect("missing group 'repo'").as_str();
+        Some(RepositoryName {
+            owner,
+            repo: repo.strip_suffix(".git").unwrap_or(repo),
+            host,
+        })
+    }
+}
+
+pub struct RepositoryStatsUpdater;
+
+impl RepositoryStatsUpdater {
+    pub(crate) fn load_repository(
+        conn: &mut Client,
+        metadata: &MetadataPackage,
+        config: Arc<Config>,
+        db: Pool,
+    ) -> Result<Option<i32>> {
+        macro_rules! return_if_ok_some {
+            ($typ:ty) => {
+                let data =
+                    Self::load_repository_inner::<$typ>(conn, metadata, config.clone(), db.clone());
+                if matches!(data, Ok(Some(_))) {
+                    return data;
+                }
+            };
+        }
+
+        // The `GitlabUpdater is a bit more permissive so better to put it at the end.
```

Comment:
Since there should be no overlap between the hosts handled by different updaters, and the repositories handled are filtered by hostname, this comment shouldn't be true.

Reply:
Not true anymore, indeed!